File: AQUA_matlab/RunAQUA2.m
%% Simulating quadratic I&F neuron with adaptation and autapse

clc
clear
close all

addpath(genpath('../HelperFunc'));

%%% RS

a=0.02; b=0.2; c=-65; d=8; % RS
vpeak=30; % spike cutoff

T=1000; h=.01; % time span and step (ms)
% I=[zeros(1,0.1*round(T/h)),500*ones(1,0.9*round(T/h))];% pulse of input DC current
% I=[zeros(1,0.1*round(T/h)),500*ones(1,0.01*round(T/h)),500*ones(1,0.89*round(T/h))];% pulse of input DC current
%I=[zeros(1,0.1*round(T/h)),4.75*ones(1,0.9*round(T/h))];% pulse of input DC current
I = [zeros(1, 40000), 4.75*ones(1, 500), zeros(1, 10000), 5.5*ones(1, 500), zeros(1, 49000)];

e=0.0; f = 0; tau = 0;
[v1,~,~,st1] = aqua2(T,h,I,vpeak,a,b,c,d,e,f,tau);
e=0.05; f = 5; tau = 5;
[v2,u2,w2,st2]= aqua2(T,h,I,vpeak,a,b,c,d,e,f,tau);
e=0.05; f = 5; tau = 0;
[v3,~,~,st3] = aqua2(T,h,I,vpeak,a,b,c,d,e,f,tau);

disp(st1(1) - st2(1))

figure(1)
subplot(3,2,1); plot(h*(1:round(T/h))/1000, v1); xlim([.008*T/1000 .03*T/1000]), ylim([-60 30]);
subplot(3,2,3); plot(h*(1:round(T/h))/1000, v2); xlim([.008*T/1000 .03*T/1000]), ylim([-60 30]);
subplot(3,2,5); plot(h*(1:round(T/h))/1000, v3); xlim([.008*T/1000 .03*T/1000]), ylim([-60 30]);
subplot(3,2,2); plot(h*(1:round(T/h))/1000, v1); xlim([.97*T/1000 T/1000]), ylim([-60 30]);
subplot(3,2,4); plot(h*(1:round(T/h))/1000, v2); xlim([.97*T/1000 T/1000]), ylim([-60 30]);
subplot(3,2,6); plot(h*(1:round(T/h))/1000, v3); xlim([.97*T/1000 T/1000]), ylim([-60 30]);

subplot(3,2,1); title('no autapse');
subplot(3,2,3); title('delayed autapse');
subplot(3,2,5); title('instantaneous autapse'); xlabel('time [sec]'); ylabel('membrane potential [mV]');
subplot(3,2,2); title('no autapse');
subplot(3,2,4); title('delayed autapse');
subplot(3,2,6); title('instantaneous autapse');

figure(2);
subplot(1,2,1); isi2 = diff(st2); plot(isi2(1:end-1),isi2(2:end),'.');
subplot(1,2,2); isi3 = diff(st3); plot(isi3(1:end-1),isi3(2:end),'.');
% subplot(1,2,1); isi2 = diff(st2); plot(isi2(end-1000+1:end-1),isi2(end-1000+2:end),'.');
% subplot(1,2,2); isi3 = diff(st3); plot(isi3(end-1000+1:end-1),isi3(end-1000+2:end),'.');

File: AQUA_python/runAQUA2.py

import numpy as np
import matplotlib.pyplot as plt
from AQUA2 import aqua2


# Parameters
a, b, c, d = 0.02, 0.2, -65, 8  # RS neuron
vpeak = 30  # Spike cutoff
T = 1000  # Total time in ms
h = 0.001  # Time step in ms
I = np.concatenate((np.zeros(int(0.01 * T / h)), 10 * np.ones(int(0.99 * T / h))))  # Input current

# Simulations
e, f, tau = 0.0, 0, 0
v1, _, _, _ = aqua2(T, h, I, vpeak, a, b, c, d, e, f, tau)

e, f, tau = 0.05, 5, 5
v2, u2, w2, st2 = aqua2(T, h, I, vpeak, a, b, c, d, e, f, tau)

e, f, tau = 0.05, 5, 0
v3, _, _, st3 = aqua2(T, h, I, vpeak, a, b, c, d, e, f, tau)

# Plot results
time = h * np.arange(0, round(T / h)) / 1000

plt.figure(figsize=(10, 6))
plt.subplot(3, 2, 1)
plt.plot(time, v1)
plt.xlim([0.008 * T / 1000, 0.06 * T / 1000])
plt.title('no autapse')

plt.subplot(3, 2, 3)
plt.plot(time, v2)
plt.xlim([0.008 * T / 1000, 0.06 * T / 1000])
plt.title('delayed autapse')

plt.subplot(3, 2, 5)
plt.plot(time, v3)
plt.xlim([0.008 * T / 1000, 0.06 * T / 1000])
plt.title('instantaneous autapse')
plt.xlabel('time [sec]')
plt.ylabel('membrane potential [mV]')

plt.subplot(3, 2, 2)
plt.plot(time, v1)
plt.xlim([0.9 * T / 1000, T / 1000])

plt.subplot(3, 2, 4)
plt.plot(time, v2)
plt.xlim([0.9 * T / 1000, T / 1000])

plt.subplot(3, 2, 6)
plt.plot(time, v3)
plt.xlim([0.9 * T / 1000, T / 1000])

plt.tight_layout()
plt.show()

# ISI Plot
plt.figure(figsize=(10, 4))
isi2 = np.diff(st2)
isi3 = np.diff(st3)

plt.subplot(1, 2, 1)
plt.plot(isi2[:-1], isi2[1:], '.')
plt.title('ISI - delayed autapse')

plt.subplot(1, 2, 2)
plt.plot(isi3[:-1], isi3[1:], '.')
plt.title('ISI - instantaneous autapse')

plt.tight_layout()
plt.show()